napari/conftest.py

```diff
@@ -429,6 +431,19 @@ def _dangling_qthreads(monkeypatch, qtbot, request):
 
     base_start = QThread.start
     thread_dict = WeakKeyDictionary()
+    base_constructor = QThread.__init__
+
+    def my_run(self):  # pragma: no cover
+        if 'coverage' in sys.modules:
+            # https://github.com/nedbat/coveragepy/issues/686#issuecomment-634932753
+            sys.settrace(threading._trace_hook)
+        self._base_run()
+
+    def my_constructor(self, *args, **kwargs):
+        base_constructor(self, *args, **kwargs)
+        self._base_run = self.run
+        self.run = partial(my_run, self)
```

Comment:
Ok I was trying to understand these... Here's my understanding:

- the names "my_run" and "my_constructor" are arbitrary — we could use more descriptive names, right?
- if so, I would propose `_init_with_trace` and `_run_with_trace`
- is assigning `base_constructor` needed? Could we not replace that line (443) with `QThread.__init__(self, *args, **kwargs)`?
- and add the following docstrings:
  - for `_init_with_trace`:
    ```
    """Constructor for QThread adding tracing for coverage measurements.

    Functions running in QThreads don't get measured by coverage.py, see
    https://github.com/nedbat/coveragepy/issues/686. Therefore, we will
    monkeypatch the constructor to add to the thread to `sys.settrace` when
    we call `run` and `coverage` is in `sys.modules`.
    """
    ```
  - for `_run_with_trace`:
    ```
    """QThread.run but adding execution to sys.settrace when measuring coverage.

    See https://github.com/nedbat/coveragepy/issues/686#issuecomment-634932753
    and `_init_with_trace`. When running QThreads during testing, we monkeypatch
    the QThread constructor and run methods with traceable equivalents.
    """
    ```

(I'd also rename `_base_run` to `_qthread_run`)

Reply:
(Is my understanding correct and if so what do you think about these suggestions @Czaki?)

Reply:
> is assigning `base_constructor` needed? Could we not replace that line (443) with `QThread.__init__(self, *args, **kwargs)`?

It may not work if more than one fixtures will try to do something with QThread constructor.